packages/core/src/actions/resolveActionsAndContext.ts

```diff
@@ -0,0 +1,210 @@
+          }
+          break;
+        case actionTypes.log:
+          resActions.push(
+            resolveLog(
+              actionObject as LogAction<TContext, TEvent>,
+              context,
+              _event
+            )
+          );
+          break;
+        case actionTypes.choose: {
+          const chooseAction = actionObject as ChooseAction<TContext, TEvent>;
+          const chosenActions = chooseAction.guards.find((condition) => {
+            const guard =
+              condition.guard &&
+              toGuardDefinition(
+                condition.guard,
+                (guardType) => machine.options.guards[guardType]
+              );
+            return (
+              !guard ||
+              evaluateGuard(guard, context, _event, currentState as any)
+            );
+          })?.actions;
+
+          if (chosenActions) {
+            toActionObjects(
+              toArray(chosenActions),
+              machine.options.actions
+            ).forEach(resolveAction);
+          }
+          break;
+        }
+
+        case actionTypes.pure:
+          const matchedActions = (actionObject as PureAction<
+            TContext,
+            TEvent
+          >).get(context, _event.data);
+
+          if (matchedActions) {
+            toActionObjects(
+              toArray(matchedActions),
+              machine.options.actions
+            ).forEach(resolveAction);
+          }
+          break;
+        case actionTypes.assign:
+          try {
+            const [nextContext, nextActions] = updateContext(
+              context,
+              _event,
+              [actionObject as AssignAction<TContext, TEvent>],
+              currentState,
+              service
+            );
+            context = nextContext;
+            resActions.push(actionObject, ...nextActions);
+          } catch (err) {
+            // Raise error.execution events for failed assign actions
+            raisedActions.push({
+              type: actionTypes.raise,
+              _event: toSCXMLEvent({
+                type: actionTypes.errorExecution,
+                error: err
+              } as any) // TODO: fix
+            });
+          }
+          break;
+        case actionTypes.stop:
+          const stopAction = resolveStop(
+            actionObject as StopActionObject,
+            context,
+            _event
+          );
+          resActions.push(stopAction);
+          break;
+        case actionTypes.forEach:
+          const action = actionObject as ForEachAction<TContext, TEvent>;
+          const array = context[action.array];
+
+          if (isArray(array)) {
+            array.forEach((item, index) => {
+              const [nextContext] = updateContext(
+                context,
+                _event,
+                [
+                  assign<TContext, TEvent>({
+                    [action.item]: item as any,
+                    [action.index]: index as any // TODO: fix
```

Comment:
Index is optional in SCXML and in here you would add an extra property under the undefined key which would raise eyebrows when people would see that while debugging